src/Migration/Sources/Csv.php

```diff
@@ -0,0 +1,338 @@
+            if (empty($fetched)) {
+                break;
+            }
+
+            $allAttributes = array_merge($allAttributes, $fetched);
+            $lastAttribute = $fetched[count($fetched) - 1];
+
+            if (count($fetched) < $batchSize) {
+                break;
+            }
+        }
+
+        $attributeTypes = [];
+        $manyToManyKeys = [];
+
+        foreach ($allAttributes as $attribute) {
+            $key = $attribute['key'];
+
+            if (
+                // Skip child-side relationships entirely
+                $attribute['type'] === Attribute::TYPE_RELATIONSHIP &&
+                ($attribute['side'] ?? '') === UtopiaDatabase::RELATION_SIDE_CHILD
+            ) {
+                continue;
+            }
+
+            $attributeTypes[$key] = $attribute['type'];
+
+            if (
+                $attribute['type'] === Attribute::TYPE_RELATIONSHIP &&
+                ($attribute['relationType'] ?? '') === 'manyToMany' &&
+                ($attribute['side'] ?? '') === 'parent'
+            ) {
+                $manyToManyKeys[] = $key;
+            }
+        }
+
+        $buffer = [];
+
+        while (($row = fgetcsv($stream)) !== false) {
+            $data = array_combine($headers, $row);
+            if ($data === false) {
+                continue;
+            }
+
+            $parsedData = $data;
+
+            foreach ($data as $key => $value) {
+                if (! isset($attributeTypes[$key])) {
+                    continue;
+                }
+
+                $type = $attributeTypes[$key];
+                $parsedValue = trim($value);
+
+                if ($parsedValue === '') {
+                    $parsedData[$key] = null;
+
+                    continue;
+                }
+
+                if (in_array($key, $manyToManyKeys, true)) {
+                    $parsedData[$key] = str_contains($parsedValue, ',')
+                        ? array_map('trim', explode(',', $parsedValue))
+                        : [$parsedValue];
+
+                    continue;
+                }
+
+                switch ($type) {
+                    case Attribute::TYPE_INTEGER:
+                        $parsedData[$key] = is_numeric($parsedValue) ? (int) $parsedValue : null;
+                        break;
+
+                    case Attribute::TYPE_FLOAT:
+                        $parsedData[$key] = is_numeric($parsedValue) ? (float) $parsedValue : null;
+                        break;
+
+                    case Attribute::TYPE_BOOLEAN:
+                        $parsedData[$key] = filter_var($parsedValue, FILTER_VALIDATE_BOOLEAN);
+                        break;
+
+                    default:
+                        break;
+                }
+            }
+
+            $permissions = [];
+            $documentId = $parsedData['$id'] ?? 'unique()';
+
+            if ($hasDocumentSecurityEnabled && isset($parsedData['$permissions'])) {
```

Comment:
We should set permissions regardless of whether documentSecurity is enabled, so that at a later time, they can be enabled without data changes

Reply:
What permission do we set on the document if its not enabled? Shouldn't the collection level permission apply anyways?

Reply:
Ahh nvm, understood. Making the change. Those are added just not enabled until later.

Reply:
Whatever was passed, or empty array if none. Collection level will apply anyway, but document level checks are skipped if document level isn't enabled. But if you enable it later, any previously added permissions (like the ones here) will apply. We use the same pattern in Appwrite for documents